man/xnvme.1
.\" Text automatically generated by txt2man
.TH untitled  "04 September 2023" "" ""
.SH NAME
\fBxnvme \fP- No short description
.SH SYNOPSIS
.nf
.fam C
\fBxnvme\fP <command> [<args>]
.fam T
.fi
.fam T
.fi
.SH DESCRIPTION
No long description
.SH COMMANDS
.TP
.B
\fBxnvme-list\fP(1)
List devices on the system
.TP
.B
\fBxnvme-enum\fP(1)
Enumerate devices on the system
.TP
.B
\fBxnvme-info\fP(1)
Retrieve derived information for given device
.TP
.B
\fBxnvme-idfy\fP(1)
Execute an User-defined Identify Command
.TP
.B
\fBxnvme-idfy-ns\fP(1)
Identify the given Namespace
.TP
.B
\fBxnvme-idfy-ctrlr\fP(1)
Identify the given Controller
.TP
.B
\fBxnvme-idfy-cs\fP(1)
Identify the Command Sets supported by the controller
.TP
.B
\fBxnvme-log\fP(1)
Retrieve and print log
.TP
.B
\fBxnvme-log-erri\fP(1)
Retrieve and print log
.TP
.B
\fBxnvme-log-health\fP(1)
Retrieve and print log
.TP
.B
\fBxnvme-log-fdp-config\fP(1)
Retrieve and print log
.TP
.B
\fBxnvme-log-ruhu\fP(1)
Retrieve and print log
.TP
.B
\fBxnvme-log-fdp-stats\fP(1)
Retrieve and print log
.TP
.B
\fBxnvme-log-fdp-events\fP(1)
Retrieve and print log
.TP
.B
\fBxnvme-feature-get\fP(1)
Execute a Get Features Command
.TP
.B
\fBxnvme-feature-set\fP(1)
Execute a Set-Features Command
.TP
.B
\fBxnvme-set-fdp-events\fP(1)
Enable or disable all events
.TP
.B
\fBxnvme-format\fP(1)
Format a NVM namespace
.TP
.B
\fBxnvme-sanitize\fP(1)
Sanitize\.\.\.
.TP
.B
\fBxnvme-fdp-ruhs\fP(1)
Retrieve and print log
.TP
.B
\fBxnvme-fdp-ruhu\fP(1)
Reclaim Unit Handle Update for a Placement Identifier
.TP
.B
\fBxnvme-pioc\fP(1)
Pass a used-defined IO Command through
.TP
.B
\fBxnvme-padc\fP(1)
Pass a user-defined ADmin Command through
.TP
.B
\fBxnvme-library-info\fP(1)
Produce information about the library
.TP
.B
\fBxnvme-dsm\fP(1)
Dataset Management
.RE
.PP

.SH OPTIONS
\fB--help\fP
Print the synopsis and exit
.SH EXAMPLES
Read the man page for each <command> or consult the command-line \fB--help\fP:
.PP
.nf
.fam C
    $ xnvme <command> --help

.fam T
.fi
.SH SEE ALSO
Full documentation at: <https://xnvme.io/>
.SH AUTHOR
Written by Simon A. F. Lund <simon.lund@samsung.com> on behalf of Samsung


CHANGELOG.rst
.. SPDX-FileCopyrightText: Samsung Electronics Co., Ltd
..
.. SPDX-License-Identifier: BSD-3-Clause

Changelog
=========

The repository is tagged using semantic versioning, e.g. ``v1.2.3``.

The ``next`` branch contains a preview of the upcoming release. The history on
``next`` is subject to change. Upon release, ``next`` is merged with ``main``
and  and tagged ``vX.Y.Z``.

Changes are described in this file in a section matching the version tag.

Known Issues
------------

See the file named ``ISSUES`` in the root of the repository.

v0.7.1
------

* API
  - Introduced memory-mapping API (libxnvme_mem.h), that is, when the user
    manages memory and thus does not want to use xnvme_buf_alloc() as doing so
    would introduce bounce-buffers.

* CLI
  - The passthru cli-interfaces now use --cdwXY arguments instead of
    "command-files", this alignes with nvme-cli and is arguably simpler

* Backends: Linux
  - Performance improvments in io_uring and io_uring_cmd backends due to the
    addition batching via delayed submission
  - Performance improvements to libaio due to the addition of the "ring-hack"

* Backends: Windows
  - Support added for writing via "ioring" on Windows, previously only batched
    reads were possible via this interface.

* Infrastructure; github-actions, docker-images, third-party etc. bump the
  reference toolchain platforms, that is
  - Debian is now Trixie, Bookworm and Bullseye.
  - Ubuntu is now Lunar, Jammy and Focal
  - Added Rocky Linux 9.2
  - Added Oracle Linux 9.
  - Dropped CentOS 7
  - Dropped CentOS Stream 8
  - Dropped openSUSE Leap 15.3 and 15.4, Tumbleweed remains

* Removal of a bunch of deprecated files and code. General *cleanup*.

v0.7.0
------

This release got a whole lot bigger than it ought to. Aiming for more
frequent releases in the future. A bunch of changes to API,
implementation, and even the license.

Regarding the license. For compatibility with GPL-licensed software.
That is, consumption of xNVMe from GPL-licences software, then xNVMe is
changed from APACHE to BSD-3-Clause. To better determine the license,
the project has aligned with the golden practice of REUSE-compliance
(https://reuse.software/).

* API
  * Refactored to provide a single "entry-point" header
  * Headers no longer include other headers (except for the entry-point)
  * Renamed the command-line api from xnvmec to xnvme_cli

* API / Command-Sets
  * Support for FDP and DSM
  * Support for Key-Value SSDs

* Backends
  * The POSIX backend is replaced by CBI -- Common Backend Implementation
  * Instead of "POSIX" then a handful of commonly useful implementations
    are provided, being "strictly" POSIX was an impractical restriction,
    such as lacking support for features which are provided on modern
    operating systems but not defined by POSIX.

* Backends: FreeBSD
  * Support for kqueue based aio

* Backends: SPDK
  * Now has options to control Admin and I/O command timeout
  * Removed patches no longer needed (available upstream)

* Backends: vfio
  * Removed as subproject and instead links "properly" with the
    system-provided libvfn

* tools
  * switched fio from "external" to "internal" / "upstream" fio engine

For additional details then please have a look at the commit-history,
the v0.7.0 related issue on GitHUB and PR.

v0.6.0
------

A handful of improvements for Windows, additional steps toward removing
third-party vendoring / bundling, addition of experimental tunable knobs for
io_uring/SQPOLL, and several improvements to testing and CI infra.

* API
  - Removed SLIST from API, although "sys/queue.h" are commonly available on
    Linux/FreeBSD, then they are not part of toolchain on Windows.

* Third-party
  - Bumped SPDK to v22.09, and with that removed mutliple out-of-tree patches
    for DPDK.

* CLI
  - The xNVMe command-line library (libxnvmec) and all the cli-tools using it
    are refactored to use common sets of command-line arguments. Along with
    this came a consistent set of CLI-arguments for admin/sync/async.

* Backends
  - ramdisk: The ramdisk got support for write-zeroes, iovec payloads and added
    to CI testing.
  - linux: support for a buffer-allocator using HUGEPAGES and tunable knobs for
    controlling the behavior of io_uring SQPOLL via environment variables.
  - windows: support for the experimental IORING Windows SPDK API and support
    for block devices (SCSI and SATA).
  - spdk: when controllers are re-used for device-handles, events are
    processed as a means to check whether the controller is still "alive"

* CI
  - scan-build now runs for each PR
  - basic tests are now running post-building testing using the RAMDISK
  - Testing of fabrics with TCP transport is now part of the setup

v0.5.0
------

A bit of expansion in the application of xNVMe with support for macOS, a
ramdisk backend, revival of Python language bindings and a refresh of the docs
on NVMe-over-Fabrics.

* API
  - Removed helpers for SGL
  - Add 'subnqn' to 'xnvme_ident', useful for NVMe-oF
  - Add 'hostnqn' to 'opts', useful for NVMe-oF
  - Add support directive-receive and write-with-directives
  - Spec adjusted for NVMe 2.0, still more work needed in this area

* Third-party
  - Bumped fio to v3.32
  - Bumped SPDK to v22.05
  - Removed liburing, now relies on on-system library instead of
    vendoring/bundling, documentation is updated to assist with library
    installation

* Tooling
  - Re-working testing using cijoe 0.9+, that is, switching to CIJOE/pytest for
    testing an CIJOE workflows for instrumentation
  - liburing is no longer bundled with xNVMe, that is, xNVMe now links with
    liburing as discovered on the system via pkg-config. This is done to avoid
    symbol collisions for applications linking or loading liburing and xNVMe.

* be:linux:async:libaio
  - When 'opts.poll_io' is set then poke() will return immmediatly and now wait
    for completions. This allows the use of trading CPU for more IOPS and lower

toolbox/cijoe-pkg-xnvme/modules/xnvme.sh
#!/usr/bin/env bash
#
# CIJOE module to verify environment variables for xNVMe testcases
#
# For more information about xNVMe, see: https://xnvme.io/
#
# For documentation of parameters / variables.
#
# xnvme.env    - Checks environment for module parameters
# xnvme.fioe   - Runs fio with the xNVMe fio IO Engine
#
xnvme.env() {
  #: "${XNVME_SHARE_ROOT:=/usr/share/xnvme}"
  export XNVME_SHARE_ROOT

  #: "${XNVME_LIB_ROOT:=/usr/lib}"
  export XNVME_LIB_ROOT

  return 0
}

#
# Uses:
# - xnvme.env
# - XNVME_URI
# - CMD_PREFIX (optional)
# - XNVME_SHARE_ROOT
# - XNVME_LIB_ROOT
#
xnvme.fioe() {
  if [[ -z "$1" || -z "$2" ]]; then
    cij.err "usage: xnvme.fioe('fio-script.fname', 'ioengine_name')"
    cij.err " or"
    cij.err "usage: xnvme.fioe('fio-script.fname', 'ioengine_name', 'section')"
    return 1
  fi

  if ! xnvme.env; then
    cij.err "xnvme:fioe: invalid xNVMe environment"
    return 1
  fi

  : "${CMD_PREFIX:=taskset -c 1 perf stat}"
  : "${FIO_BIN:=fio}"

  local script_fname=$1
  local ioengine_name=$2
  local section=$3
  local output_fpath=$4

  local jobfile="${script_fname}"
  if [[ -v XNVME_SHARE_ROOT && -n "${XNVME_SHARE_ROOT}" ]]; then
    jobfile="${XNVME_SHARE_ROOT}/${jobfile}"
  fi
  local _cmd

  if ! cij.cmd "[[ -f \"${jobfile}\" ]]"; then
    cij.err "xnvme:fioe: '${jobfile}' does not exist!"
    #return 1
  fi

  _cmd="${CMD_PREFIX} ${FIO_BIN} ${jobfile}"
  if [[ -v FIO_BS ]]; then
    _cmd="FIO_BS=${FIO_BS} ${_cmd}"
  fi
  if [[ -v FIO_RW ]]; then
    _cmd="FIO_RW=${FIO_RW} ${_cmd}"
  fi
  if [[ -v FIO_IODEPTH ]]; then
    _cmd="FIO_IODEPTH=${FIO_IODEPTH} ${_cmd}"
  fi
  _cmd="${_cmd} --section=${section}"

  # Add the special-sauce for the external xNVMe fio engine
  if [[ "$ioengine_name" == *"xnvme"* ]]; then
    : "${XNVME_URI:?Must be set and non-empty}"
    : "${XNVME_ASYNC:?Must be set and non-empty}"
    : "${XNVME_SYNC:?Must be set and non-empty}"
    : "${XNVME_ADMIN:?Must be set and non-empty}"

    local _fioe_so
    if [[ -v XNVME_BE && "${XNVME_BE}" == "windows" ]]; then
      _fioe_so="${XNVME_LIB_ROOT}libxnvme-fio-engine.dll"
    else
      _fioe_so="${XNVME_LIB_ROOT}/libxnvme-fio-engine.so"
    fi
    local _fioe_uri=${XNVME_URI//:/\\\\:}

    if ! cij.cmd "[[ -f \"${_fioe_so}\" ]]"; then
      cij.err "xnvme:fioe: '${_fioe_so}' does not exist!"
      #return 1
    fi

    _cmd="${_cmd} --ioengine=\"external:${_fioe_so}\""
    _cmd="${_cmd} --xnvme_async=${XNVME_ASYNC}"
    _cmd="${_cmd} --xnvme_sync=${XNVME_SYNC}"
    _cmd="${_cmd} --xnvme_admin=${XNVME_ADMIN}"
    _cmd="${_cmd} --xnvme_dev_nsid=${XNVME_DEV_NSID}"
    _cmd="${_cmd} --filename=\"${_fioe_uri}\""

  # Add the special-sauce for the external SPDK io-engine spdk_nvme
  elif [[ "$ioengine_name" == "spdk_nvme" ]]; then

    local _traddr=${PCI_DEV_NAME//:/.};

    _cmd="LD_PRELOAD=${SPDK_FIOE_ROOT}/${ioengine_name} ${_cmd}"
    _cmd="${_cmd} --ioengine=spdk"
    _cmd="${_cmd} --filename=\"trtype=PCIe traddr=${_traddr} ns=${NVME_NSID}\""

  # Add the special-sauce for the external SPDK/nvme_bdev
  elif [[ "$ioengine_name" == "spdk_bdev" ]]; then
    # Produce SPDK config-file
    #echo "[Nvme]" > /tmp/spdk.bdev.conf
    #echo "  TransportID \"trtype:PCIe traddr:${PCI_DEV_NAME}\" Nvme0" >> /tmp/spdk.bdev.conf
    #ssh.push /tmp/spdk.bdev.conf /opt/aux/spdk.bdev.conf
    cij.emph "engine: spdk_bdev engine... [$XNVME_FOO] --- [$SPDK_JSON_CONF]"

    local _filename=${SPDK_FIOE_FILENAME//:/.};
    local _spdk_json_conf=${SPDK_JSON_CONF}

    _cmd="LD_PRELOAD=${SPDK_FIOE_ROOT}/${ioengine_name} ${_cmd}"
    _cmd="${_cmd} --ioengine=${ioengine_name}"
    _cmd="${_cmd} --spdk_json_conf=\"${_spdk_json_conf}\""
    _cmd="${_cmd} --filename=\"${_filename}\""

  # Add the not-so-special-sauce for built-in io-engines
  else
    _cmd="${_cmd} --ioengine=${ioengine_name}"
    _cmd="${_cmd} --filename=\"${XNVME_URI}\""
  fi

  if [[ -v FIO_AUX ]]; then
    _cmd="${_cmd} ${FIO_AUX}"
  fi

  local _trgt_fname="/tmp/fio-output.txt"

  _cmd="${_cmd} --output-format=normal,json --output=${_trgt_fname}"

  # Now run that fio tester!
  if ! cij.cmd "${_cmd}"; then
    cij.err "xnvme.fioe: error running fio"

    ssh.pull "${_trgt_fname}" "${output_fpath}"
    return 1
  fi

  # Download the output
  if ! ssh.pull "${_trgt_fname}" "${output_fpath}"; then
    cij.err "xnvme.fioe: failed pulling down fio output-file"
    return 0
  fi

  # Dump it to stdout for prosperity
  cat "${output_fpath}"

  return 0
}
